orchagent/vnetorch.cpp

```diff
@@ -960,62 +1098,110 @@ bool VNetRouteOrch::doRouteTask<VNetVrfObject>(const string& vnet, IpPrefix& ipP
             {
                 SWSS_LOG_ERROR("Route add/update failed for %s, vr_id '0x%" PRIx64, ipPrefix.to_string().c_str(), vr_id);
                 /* Clean up the newly created next hop group entry */
-                if (nexthops.getSize() > 1)
+                if (active_nhg.getSize() > 1)
                 {
-                    removeNextHopGroup(vnet, nexthops, vrf_obj);
+                    removeNextHopGroup(vnet, active_nhg, vrf_obj);
                 }
                 return false;
             }
         }
-
-        if (it_route != syncd_tunnel_routes_[vnet].end() && it_route->second != nexthops)
+        bool route_updated = false;
+        bool priority_route_updated = false;
+        if (it_route != syncd_tunnel_routes_[vnet].end() &&
+            ((monitoring == "" && it_route->second.nhg_key != nexthops) ||
+            (monitoring == "custom" && (it_route->second.primary != nexthops || it_route->second.secondary != nexthops_secondary))))
         {
-            // In case of updating an existing route, decrease the reference count for the previous nexthop group if not same as new nhg
-            NextHopGroupKey nhg = it_route->second;
-            if (--syncd_nexthop_groups_[vnet][nhg].ref_count == 0)
+            route_updated = true;
+            NextHopGroupKey nhg = it_route->second.nhg_key;
+            if (monitoring == "custom")
             {
-                if (nhg.getSize() > 1)
+                // if the previously active NHG is same as the newly created active NHG.case of primary secondary swap or
+                //when primary is active and secondary is changed or vice versa. In these cases we dont remove the NHG
+                // but only remove the monitors for the set which has changed.
+                if (it_route->second.primary != nexthops)
                 {
-                    removeNextHopGroup(vnet, nhg, vrf_obj);
```

Comment:
Where is removeNextGroup called for the "custom" case after removing here?

Reply:
its done in updateVnetTunnelCustomMonitor because we are setting priority_route_updated